packages/swing-store-lmdb/src/sqlStreamStore.js

```diff
@@ -0,0 +1,148 @@
+ */
+function insistStreamPosition(position) {
+  assert.typeof(position, 'object');
+  assert(position);
+  assert.typeof(position.itemCount, 'number');
+  assert(position.itemCount >= 0);
+  return undefined;
+}
+
+/**
+ * @param {string} dbDir
+ * @param {{ sqlite3?: typeof import('better-sqlite3') }} [io]
+ */
+export function sqlStreamStore(dbDir, io) {
+  const { sqlite3 = sqlite3ambient } = io || {};
+  const filePath = `${dbDir}/streams.db`; // ISSUE: no path.join?
+  const db = sqlite3(
+    filePath,
+    // { verbose: console.log },
+  );
+  db.exec(`
+    create table if not exists streamItem (
+      streamName text,
+      position integer,
+      item text,
+      primary key (streamName, position)
+    )
+  `);
+
+  const streamStatus = new Map();
+
+  /**
+   * @param {string} streamName
+   * @param {StreamPosition} startPosition
+   * @param {StreamPosition} endPosition
+   */
+  function readStream(streamName, startPosition, endPosition) {
+    insistStreamName(streamName);
+    // TODO: enforce exclusive access? DB transactions are isolated.
+    // assert(
+    //   !streamStatus.get(streamName),
+    //   X`can't read stream ${q(streamName)} because it's already in use`,
+    // );
+    insistStreamPosition(startPosition);
+    insistStreamPosition(endPosition);
+    assert(
+      startPosition.itemCount <= endPosition.itemCount,
+      X`${q(startPosition.itemCount)} <= ${q(endPosition.itemCount)}}`,
+    );
+
+    function* reader() {
+      const query = db.prepare(`
+        select item
+        from streamItem
+        where streamName = ? and position >= ? and position < ?
+        order by position
+      `);
+      for (const { item } of query.iterate(
+        streamName,
+        startPosition.itemCount,
+        endPosition.itemCount,
+      )) {
+        assert(
+          streamStatus.get(streamName) === 'reading',
+          X`can't read stream ${q(streamName)}, it's been closed`,
+        );
+        yield item;
+      }
+      streamStatus.delete(streamName);
+    }
+
+    assert(
+      !streamStatus.has(streamName),
+      X`can't read stream ${q(streamName)} because it's already in use`,
+    );
+
+    if (startPosition.itemCount === endPosition.itemCount) {
+      return [];
+    }
+
+    streamStatus.set(streamName, 'reading');
+
+    return reader();
+  }
+
+  /**
+   * @param {string} streamName
+   * @param {string} item
+   * @param {StreamPosition} position
+   */
+  const writeStreamItem = (streamName, item, position) => {
+    insistStreamName(streamName);
+    insistStreamPosition(position);
+
+    assert(
+      [undefined, 'writing'].includes(streamStatus.get(streamName)),
+      X`can't write stream ${q(streamName)} because it's already in use`,
+    );
+    streamStatus.set(streamName, 'writing');
+
+    db.prepare(
+      `
+      insert into streamItem (streamName, item, position)
+        values (?, ?, ?)
+        on conflict(streamName, position) do update set item = ?
```

Comment:
Am I correct in understanding that the `on conflict...` verbiage handles the case where we're writing over the top of previously committed but abandoned transcript entries?